libcloud/compute/drivers/ecs.py

```diff
@@ -0,0 +1,1390 @@
+        self.available_instance_types = available_instance_types
+        self.available_disk_categories = available_disk_categories
+
+    def __repr__(self):
+        return ('<ECSZone: id=%s, name=%s, driver=%s>' %
+                (self.id, self.name, self.driver))
+
+
+class InternetChargeType(object):
+    """
+    Internet connection billing types for Aliyun Nodes.
+    """
+    BY_BANDWIDTH = 'PayByBandwidth'
+    BY_TRAFFIC = 'PayByTraffic'
+
+
+class DiskCategory(object):
+    """
+    Enum defined disk types supported by Aliyun system and data disks.
+    """
+    CLOUD = 'cloud'
+    CLOUD_EFFICIENCY = 'cloud_efficiency'
+    CLOUD_SSD = 'cloud_ssd'
+    EPHEMERAL_SSD = 'ephemeral_ssd'
+
+
+class Pagination(object):
+    """
+    Pagination used to describe the multiple pages results.
+    """
+    def __init__(self, total, size, current):
+        """
+        Create a pagination.
+
+        :param total: the total count of the results
+        :param size: the page size of each page
+        :param current: the current page number, 1-based
+        """
+        self.total = total
+        self.size = size
+        self.current = current
+
+    def next(self):
+        """
+        Switch to the next page.
+        :return: the new pagination or None when no more page
+        :rtype: ``Pagination``
+        """
+        if self.total is None or (self.size * self.current >= self.total):
+            return None
+        self.current += 1
+        return self
+
+    def to_dict(self):
+        return {'PageNumber': self.current,
+                'PageSize': self.size}
+
+    def __repr__(self):
+        return ('<Pagination total=%d, size=%d, current page=%d>' %
+                (self.total, self.size, self.current))
+
+
+class ECSDriver(NodeDriver):
+    """
+    Aliyun ECS node driver.
+
+    Used for Aliyun ECS service.
+    """
+
+    name = 'Aliyun ECS'
+    website = 'https://www.aliyun.com/product/ecs'
+    connectionCls = ECSConnection
+    features = {'create_node': ['password']}
+    namespace = None
+    path = '/'
+
+    internet_charge_types = InternetChargeType
+    disk_categories = DiskCategory
+
+    NODE_STATE_MAPPING = {
+        'Starting': NodeState.PENDING,
+        'Running': NodeState.RUNNING,
+        'Stopping': NodeState.PENDING,
+        'Stopped': NodeState.STOPPED
+    }
+
+    VOLUME_STATE_MAPPING = {
+        'In_use': StorageVolumeState.INUSE,
+        'Available': StorageVolumeState.AVAILABLE,
+        'Attaching': StorageVolumeState.ATTACHING,
+        'Detaching': StorageVolumeState.INUSE,
+        'Creating': StorageVolumeState.CREATING,
+        'ReIniting': StorageVolumeState.CREATING}
+
+    SNAPSHOT_STATE_MAPPING = {
+        'progressing': VolumeSnapshotState.CREATING,
+        'accomplished': VolumeSnapshotState.AVAILABLE,
+        'failed': VolumeSnapshotState.ERROR}
+
+    def list_nodes(self, ex_node_ids=[], ex_filters=None):
```

Comment:
Minor thing - please don't default to a mutable type (list).

Rather to `ex_node_ids=None` and inside method body: `ex_node_ids = ex_node_ids or []`